KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,389 @@
+- Registered, Approved → Initialized, when `reset()` is called.
+
+All other status transitions are not possible.
+
+#### Structs
+
+The smart contract will have the following structs:
+
+- `Retired`: to represent the details of retired and approver addresses.
+  - Retired represents the previous fund addresses such as KGF or KIR
+  - Approver represents the admin of the retired address. Approver addresses are stored to track the approvals received for asset transfer
+- `Newbie`: to represent newbies and their fund allocation.
+  - Newbie represents the 'rebalanced fund' such as KCF or KFF
+
+#### Storage
+
+The smart contract will have the following storage variables:
+
+- `retirees`: array of `Retired` struct.
+- `newbies`: array of `Newbie` struct.
+- `status`: current status of the contract.
+- `rebalanceBlockNumber`: the target block number of the execution of rebalancing.
+- `memo`: result of the treasury fund rebalance. eg : `Treasury Rebalanced Successfully`
+
+#### Modifiers
+
+The smart contract will have the following modifier:
+
+- `onlyAtStatus`: to restrict access to certain functions based on the current status of the contract. If the status is not the same with the given status, it reverts.
+
+#### Constructor
+
+The smart contract will have the following constructor:
+
+- `constructor`: to initialize the contract with the target block number of the execution of rebalance.
+
+#### State Changing Functions
+
+The smart contract will have the following state changing functions:
+
+- `registerRetired`: to register retired details. Retired stores the retired address and approvers
+- `removeRetired`: to remove retired details from the array.
+- `registerNewbie`: to register newbie address and its fund distribution.
+- `removeNewbie`: to remove newbie details from the array.
+- `finalizeRegistration`: sets the status to Registered only executed by owner. After this stage, registrations will be restricted.
+- `approve`: to approve a retiredAddress by the admin.
+- `finalizeApproval`: sets the status to Approved. After this stage, approvals will be restricted.
+  </br>
+  **Conditions**
+  - every retiredAddress must be approved
+  - min required admin’s approval is done for retired contract address.
+  - sum of retired’s balance is greater than sum of the newbies' amount
+    </br>
+- `reset`: resets all storage values to empty objects except rebalanceBlockNumber. It can only be called during Initialization, Registration and Approved status.
+- `finalizeContract`: to record the execution result of the rebalance and finalize the contract. The storage data cannot be modified after this stage.
+
+#### Fallback Function
+
+The smart contract will have a fallback function to revert any payments. Since its a treasury contract, it should not accept any payments nor it should allow any withdrawal.
+
+### Core Logic Overview
+
+TBU - describe how to read contract data and execute rebalancing in Klaytn node
```

Comment:
To enable treasury fund rebalancing, a Klaytn node should specify a deployed smart contract address and a target block number on the node configuration like hardfork items. At the configured block number, a Klaytn node reads registered information from the smart contract and executes treasury fund rebalancing.

#### Chain Configuration
ChainConfig introduces the following fields for treasury fund rebalancing. The configurations are used to trigger rebalancing and find the contract storing rebalancing information. All nodes in the network should update `genesis.json` configuration with the same value as if updating a hardfork block number. The configuration values for Baobab and Cypress networks can be hard coded on the client source code. 

- `kip103CompatibleBlock`: Treasury fund rebalancing executing block which is the same as `rebalanceBlockNumber` of the smart contract
- `kip103ContractAddress`: The address of the treasury fund rebalancing contract

#### Validation
Before executing rebalancing, registered values on the smart contract are validated to confirm whether the owners of the retired account agree on the rebalancing and whether the rebalancing doesn't cause inflation of KLAY. The following should be confirmed before the execution. 

- `kip103ContractAddress` == `contract.rebalanceBlockNumber`: Ensure the owner of the retired account agree on this timing
-  `contract.status` == `Approved`: Confirm the status of the contract is ready to rebalance 
- `contract.checkRetiredsApproved()`: Double-check the agreement of the retired accounts' owners at the execution time since the ownership is replaceable
-  `totalRetiredAmount >= totalNewbieAmount`: Ensure the rebalancing doesn't issue any KLAY

#### Execution
The rebalancing is executed at the last of the block processing process. In other words, it is executed after processing all transactions of the block and distributing block rewards. If a retired account is one of the receiver of the block reward, the amount also will be used for rebalancing. All KLAY newbies have before rebalancing will be burnt as well. After rebalancing, the remaining KLAY of retired accounts will be burnt. The following is the new balancing allocating logic including burn. 

```
for addr := range Retireds {
	state.SetBalance(addr, 0)
}

for addr, balance := Newbie {
	// if newbie has KLAY before the allocation, it will be burnt 
	currentBalance := state.GetBalance(addr) 
	Burnt = Burnt + currentBalance

	state.SetBalance(addr, balance)
}
```

#### Result
The execution result of treasury fund rebalancing will be printed as an INFO-level log on each node. The owner of the treasury rebalancing contract is supposed to update the log on the smart contract as `memo` finalizing the status of the contract. Anyone can read and verify the rebalancing result by interacting with the smart contract. The data type of `memo` is byte array containing marshaled json data. Refer to the following format and example if you want to parse it. 


```Format
{
  "retired": {"0xRetiredAddress1":[removed balance in PEB], "0xRetiredAddress2":[removed balance in PEB], ... },
  "newbie" {"0xNewbieAddress1":[new allocated balance in PEB], "0xNewbieAddress2":[new allocated balance in PEB], ... },,
  "burnt":[bunt amount in PEB],
  "success":true
}

```

```Example
memo="{
  "retired": {"0x38138d89c321b3b5f421e9452b69cf29e4380bae":117000000000000000000000000000000000000, "0x30208f32c70e8b53a67ea171c8720cbfe32888ff":117000000000000000000000000000000000000},
  "newbie" {"0x0a33a1b99bd67a7189573dd74de80293afdf969a":22500000000000000000000000000000000000, "0xd9de2697000c3665e9c5a71e1bf52aaa44507cc0":22500000000000000000000000000000000000},
  "burnt":72000000000000000000000000000000000000,
  "success":true
}
```

Reply:
@aidan-kwon How about below memo format 
```
{
  "retirees": [
    {
      "retired": "0x38138d89c321b3b5f421e9452b69cf29e4380bae",
      "balance": 20000000000000000000
    },
    {
      "retired": "0x0a33a1b99bd67a7189573dd74de80293afdf969a",
      "balance": 20000000000000000000
    }
  ],
  "newbies": [
    {
      "newbie": "0x38138d89c321b3b5f421e9452b69cf29e4380bae",
      "fundAllocated": 10000000000000000000
    },
    {
      "newbie": "0x0a33a1b99bd67a7189573dd74de80293afdf969a",
      "fundAllocated": 10000000000000000000
    }
  ],
  "burnt": 7.2e+37,
  "success": true
}
```